plugin/pkg/admission/limitranger/admission.go

```diff
@@ -161,11 +161,11 @@ func (l *LimitRanger) GetLimitRanges(a admission.Attributes) ([]*corev1.LimitRan
 
 	// if there are no items held in our indexer, check our live-lookup LRU, if that misses, do the live lookup to prime it.
 	if len(items) == 0 {
-		lruItemObj, ok := l.liveLookupCache.Get(a.GetNamespace())
-		if !ok || lruItemObj.(liveLookupEntry).expiry.Before(time.Now()) {
-			// Fixed: #22422
-			// use singleflight to alleviate simultaneous calls to
-			lruItemObj, err, _ = l.group.Do(a.GetNamespace(), func() (interface{}, error) {
+		// Fixed: #22422
+		// use singleflight to alleviate simultaneous calls to
+		lruItemObj, err, _ := l.group.Do(a.GetNamespace(), func() (interface{}, error) {
```

Comment:
why do we need to change the implementation to make cache accesses go through a singleflight to deflake the test?

Reply:
Do call after reading lru, under normal circumstances, this process is very fast, as long as no lru is read, it will directly to singleflight, that will be singleflight, and when the test is more concentrated, what is more frequent use of resources, it will Do, he has a certain time delay, in this time delay process, the last singleflight.

Modifying the implementation is a more feasible way to get the test to pass steadily without increasing the sleep case

Reply:
I don't really follow, can you explain the problem in a terms of calls and timestamps

t0 - call A1 arrives and checks cache
t1 - Do(A1) start
t2 - call A2 arrives and checks cache
t3 - A2 joins A1
t4 - DO(A1) finish and answer boths A1 and A2

Reply:
t0 - call A1 arrives and checks cache
t1 - Do(A1) start
t2 - call A2 arrives and checks cache
t3 - A2 joins A1
t4 - call A3 arrives and checks cache
t5 - DO(A1) finish and answer boths A1 and A2
t6 - Do(A3) start

Reply:
the problem is that the test deadlocks on 

       /usr/local/google/home/aojea/src/kubernetes/plugin/pkg/admission/limitranger/admission_test.go:934 +0x1854

https://github.com/kubernetes/kubernetes/blob/5bc094d2f5480ebf9df718f00d1f9ebd448c6577/plugin/pkg/admission/limitranger/admission_test.go#L929-L936

you can see the stack running with ` stress ./limitranger.test -test.run ^TestLimitRanger_GetLimitRangesFixed22422$ -test.timeout 32s`, it panics in less than a minute

The problem is that before unhold, we need to be sure all the goroutines are "captured" in the syncflight group, just adding a small sleep the test doesn´t flake
```
diff --git a/plugin/pkg/admission/limitranger/admission_test.go b/plugin/pkg/admission/limitranger/admission_test.go
index 01a83816037..e1b5e002fac 100644
--- a/plugin/pkg/admission/limitranger/admission_test.go
+++ b/plugin/pkg/admission/limitranger/admission_test.go
@@ -926,11 +926,10 @@ func TestLimitRanger_GetLimitRangesFixed22422(t *testing.T) {
                        }
                }()
        }
+       time.Sleep(1 * time.Second)
        // unhold all the calls with the same namespace handler.GetLimitRanges(attributes) calls, that have to be aggregated
        unhold <- struct{}{}
-       go func() {
-               unhold <- struct{}{}
-       }()
+       unhold <- struct{}{}

        // and here we wait for all the goroutines
        wg.Wait()
````

```
 stress ./limitranger.test -test.run ^TestLimitRanger_GetLimitRangesFixed22422$ -test.timeout 32s
5s: 192 runs so far, 0 failures
10s: 386 runs so far, 0 failures
15s: 624 runs so far, 0 failures
20s: 828 runs so far, 0 failures
25s: 1056 runs so far, 0 failures
30s: 1274 runs so far, 0 failures
35s: 1495 runs so far, 0 failures
40s: 1713 runs so far, 0 failures
45s: 1935 runs so far, 0 failures
50s: 2162 runs so far, 0 failures
55s: 2381 runs so far, 0 failures
1m0s: 2603 runs so far, 0 failures
1m5s: 2819 runs so far, 0 failures
1m10s: 3045 runs so far, 0 failures
1m15s: 3267 runs so far, 0 failures
1m20s: 3486 runs so far, 0 failures
1m25s: 3711 runs so far, 0 failures
1m30s: 3927 runs so far, 0 failures
1m35s: 4153 runs so far, 0 failures
1m40s: 4374 runs so far, 0 failures
1m45s: 4594 runs so far, 0 failures
1m50s: 4819 runs so far, 0 failures
1m55s: 5034 runs so far, 0 failures
2m0s: 5262 runs so far, 0 failures
2m5s: 5482 runs so far, 0 failures
2m10s: 5700 runs so far, 0 failures
2m15s: 5927 runs so far, 0 failures
2m20s: 6143 runs so far, 0 failures
2m25s: 6369 runs so far, 0 failures
2m30s: 6590 runs so far, 0 failures
2m35s: 6810 runs so far, 0 failures
2m40s: 7034 runs so far, 0 failures
2m45s: 7253 runs so far, 0 failures
2m50s: 7476 runs so far, 0 failures
2m55s: 7698 runs so far, 0 failures
```

Reply:
> t0 - call A1 arrives and checks cache t1 - Do(A1) start t2 - call A2 arrives and checks cache t3 - A2 joins A1 t4 - call A3 arrives and checks cache t5 - Do(A1) finish and answer boths A1 and A2 t6 - Do(A3) start // DO(A1) has already returned and cannot join, so it has to open a new call

Do(A3) 

After execution, it blocks because there is no value in chan